File: CAPI/API/src/Communication.cpp
#include "Communication.h"
#include "utils.hpp"
#include "structures.h"
#include <thread>
#include <mutex>
#include <condition_variable>

using grpc::ClientContext;

Communication::Communication(std::string sIP, std::string sPort)
{
    std::string aim = sIP + ':' + sPort;
    auto channel = grpc::CreateChannel(aim, grpc::InsecureChannelCredentials());
    THUAI6Stub = protobuf::AvailableService::NewStub(channel);
}

bool Communication::Move(int64_t time, double angle, int64_t playerID)
{
    protobuf::MoveRes moveResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufMove(time, angle, playerID);
    std::cout << "Move request sent" << std::endl;
    auto status = THUAI6Stub->Move(&context, request, &moveResult);
    if (status.ok())
        return moveResult.act_success();
    else
        return false;
}

bool Communication::PickProp(THUAI6::PropType prop, int64_t playerID)
{
    protobuf::BoolRes pickPropResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufPick(prop, playerID);
    auto status = THUAI6Stub->PickProp(&context, request, &pickPropResult);
    if (status.ok())
        return pickPropResult.act_success();
    else
        return false;
}

bool Communication::UseProp(int64_t playerID)
{
    protobuf::BoolRes usePropResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->UseProp(&context, request, &usePropResult);
    if (status.ok())
        return usePropResult.act_success();
    else
        return false;
}

bool Communication::UseSkill(int64_t playerID)
{
    protobuf::BoolRes useSkillResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->UseSkill(&context, request, &useSkillResult);
    if (status.ok())
        return useSkillResult.act_success();
    else
        return false;
}

bool Communication::SendMessage(int64_t toID, std::string message, int64_t playerID)
{
    protobuf::BoolRes sendMessageResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufSend(message, toID, playerID);
    auto status = THUAI6Stub->SendMessage(&context, request, &sendMessageResult);
    if (status.ok())
        return sendMessageResult.act_success();
    else
        return false;
}

bool Communication::Graduate(int64_t playerID)
{
    protobuf::BoolRes graduateResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->Graduate(&context, request, &graduateResult);
    if (status.ok())
        return graduateResult.act_success();
    else
        return false;
}

bool Communication::StartLearning(int64_t playerID)
{
    protobuf::BoolRes startLearningResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->StartLearning(&context, request, &startLearningResult);
    if (status.ok())
        return startLearningResult.act_success();
    else
        return false;
}

bool Communication::EndLearning(int64_t playerID)
{
    protobuf::BoolRes endLearningResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->EndLearning(&context, request, &endLearningResult);
    if (status.ok())
        return endLearningResult.act_success();
    else
        return false;
}

bool Communication::StartHelpMate(int64_t playerID)
{
    protobuf::BoolRes saveStudentResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->StartHelpMate(&context, request, &saveStudentResult);
    if (status.ok())
        return saveStudentResult.act_success();
    else
        return false;
}

bool Communication::EndHelpMate(int64_t playerID)
{
    protobuf::BoolRes saveStudentResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->EndHelpMate(&context, request, &saveStudentResult);
    if (status.ok())
        return saveStudentResult.act_success();
    else
        return false;
}

bool Communication::Trick(double angle, int64_t playerID)
{
    protobuf::BoolRes trickResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufTrick(angle, playerID);
    auto status = THUAI6Stub->Trick(&context, request, &trickResult);
    if (status.ok())
        return trickResult.act_success();
    else
        return false;
}

bool Communication::StartExam(int64_t playerID)
{
    protobuf::BoolRes startExamResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->StartExam(&context, request, &startExamResult);
    if (status.ok())
        return startExamResult.act_success();
    else
        return false;
}

bool Communication::EndExam(int64_t playerID)
{
    protobuf::BoolRes endExamResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->EndExam(&context, request, &endExamResult);
    if (status.ok())
        return endExamResult.act_success();
    else
        return false;
}

bool Communication::MakeFail(int64_t playerID)
{
    protobuf::BoolRes makeFailResult;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->MakeFail(&context, request, &makeFailResult);
    if (status.ok())
        return makeFailResult.act_success();
    else
        return false;
}

bool Communication::TryConnection(int64_t playerID)
{
    protobuf::BoolRes reply;
    ClientContext context;
    auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
    auto status = THUAI6Stub->TryConnection(&context, request, &reply);
    if (status.ok())
    {
        std::cout << "Connection success!" << std::endl;
        return true;
    }
    else
        return false;
}

protobuf::MessageToClient Communication::GetMessage2Client()
{
    std::unique_lock<std::mutex> lock(mtxMessage);
    cvMessage.wait(lock, [this]()
                   { return haveNewMessage; });
    haveNewMessage = false;
    return message2Client;
}

std::optional<std::pair<int64_t, std::string>> Communication::GetMessage()
{
    return messageQueue.tryPop();
}

bool Communication::HaveMessage()
{
    return !messageQueue.empty();
}

void Communication::ReadMessage(int64_t playerID)
{
    auto tRead = [=]()
    {
        auto request = THUAI62Proto::THUAI62ProtobufID(playerID);
        ClientContext context;
        protobuf::MsgRes messageReceived;
        auto reader = THUAI6Stub->GetMessage(&context, request);
        while (reader->Read(&messageReceived))
        {
            messageQueue.emplace(messageReceived.from_player_id(), messageReceived.message_received());
        }
    };
    std::thread(tRead).detach();
}

void Communication::AddPlayer(int64_t playerID, THUAI6::PlayerType playerType, THUAI6::StudentType studentType, THUAI6::TrickerType trickerType)
{
    auto tMessage = [=]()
    {
        protobuf::PlayerMsg playerMsg = THUAI62Proto::THUAI62ProtobufPlayer(playerID, playerType, studentType, trickerType);
        grpc::ClientContext context;
        auto MessageReader = THUAI6Stub->AddPlayer(&context, playerMsg);

        while (MessageReader->Read(&message2Client))
        {
            {
                std::lock_guard<std::mutex> lock(mtxMessage);
                haveNewMessage = true;
            }
            cvMessage.notify_one();
        }
    };
    std::thread(tMessage).detach();
}

File: PyAPI/API/Communication.py
from queue import Queue
import grpc
import threading
import proto.Message2Clients_pb2 as Message2Clients
import proto.Message2Server_pb2 as Message2Server
import proto.MessageType_pb2 as MessageType
import proto.Services_pb2_grpc as Services
from logic import Logic
from Interface import IErrorHandler
from utils import THUAI62Proto
import structures as THUAI6


# 使用gRPC的异步来减少通信对于选手而言损失的时间，而gRPC的return值有result()方法，故若连接错误时也应当返回一个具有result()方法的对象，使用此处的ErrorHandler类来实现
class BoolErrorHandler(IErrorHandler):
    @staticmethod
    def result():
        return False


class Communication:

    __THUAI6Stub: Services.AvailableServiceStub
    __haveNewMessage: bool
    __message2Client: Message2Clients.MessageToClient
    __messageQueue: Queue  # Python的Queue是线程安全的，故无须自己实现Queue

    def __init__(self, sIP: str, sPort: str):
        aim = sIP + ':' + sPort
        channel = grpc.insecure_channel(aim)
        self.__THUAI6Stub = Services.AvailableServiceStub(channel)
        self.__haveNewMessage = False
        self.__messageQueue = Queue()

    def Move(self, time: int, angle: float, playerID: int) -> bool:
        try:
            moveResult = self.__THUAI6Stub.Move(
                THUAI62Proto.THUAI62ProtobufMove(time, angle, playerID))
        except grpc.RpcError as e:
            return False
        else:
            return moveResult.act_success

    def PickProp(self, propType: THUAI6.PropType, playerID: int) -> bool:
        try:
            pickResult = self.__THUAI6Stub.PickProp(
                THUAI62Proto.THUAI62ProtobufPick(propType, playerID))
        except grpc.RpcError as e:
            return False
        else:
            return pickResult.act_success

    def UseProp(self, playerID: int):
        try:
            useResult = self.__THUAI6Stub.UseProp(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return useResult.act_success

    def UseSkill(self, playerID: int) -> bool:
        try:
            useResult = self.__THUAI6Stub.UseSkill(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return useResult.act_success

    def SendMessage(self, toID: int, message: str, playerID: int) -> bool:
        try:
            sendResult = self.__THUAI6Stub.SendMessage(
                THUAI62Proto.THUAI62ProtobufSend(message, toID, playerID))
        except grpc.RpcError as e:
            return False
        else:
            return sendResult.act_success

    def HaveMessage(self) -> bool:
        return not self.__messageQueue.empty()

    def GetMessage(self) -> tuple[int, str]:
        try:
            message = self.__messageQueue.get_nowait()
        except Exception as e:
            return -1, ''
        else:
            return message

    def ReadMessage(self, playerID: int) -> None:
        def tRead():
            try:
                for msg in self.__THUAI6Stub.GetMessage(
                        THUAI62Proto.THUAI62ProtobufID(playerID)):
                    self.__messageQueue.put(
                        (msg.from_player_id, msg.message_received))
            except grpc.RpcError as e:
                return

        threading.Thread(target=tRead).start()

    def Escape(self, playerID: int) -> bool:
        try:
            escapeResult = self.__THUAI6Stub.Escape(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return escapeResult.act_success

    def StartFixMachine(self, playerID: int) -> bool:
        try:
            fixResult = self.__THUAI6Stub.StartFixMachine(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return fixResult.act_success

    def EndFixMachine(self, playerID: int) -> bool:
        try:
            fixResult = self.__THUAI6Stub.EndFixMachine(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return fixResult.act_success

    def StartSaveHuman(self, playerID: int) -> bool:
        try:
            saveResult = self.__THUAI6Stub.StartSaveHuman(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return saveResult.act_success

    def EndSaveHuman(self, playerID: int) -> bool:
        try:
            saveResult = self.__THUAI6Stub.EndSaveHuman(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return saveResult.act_success

    def Attack(self, angle: float, playerID: int) -> bool:
        try:
            attackResult = self.__THUAI6Stub.Attack(
                THUAI62Proto.THUAI62ProtobufAttack(angle, playerID))
        except grpc.RpcError as e:
            return False
        else:
            return attackResult.act_success

    def CarryHuman(self, playerID: int) -> bool:
        try:
            carryResult = self.__THUAI6Stub.CarryHuman(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return carryResult.act_success

    def ReleaseHuman(self, playerID: int) -> bool:
        try:
            releaseResult = self.__THUAI6Stub.ReleaseHuman(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return releaseResult.act_success

    def HangHuman(self, playerID: int) -> bool:
        try:
            hangResult = self.__THUAI6Stub.HangHuman(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return hangResult.act_success

    def TryConnection(self, playerID: int) -> bool:
        try:
            connectResult = self.__THUAI6Stub.TryConnection(
                THUAI62Proto.THUAI62ProtobufID(playerID))
        except grpc.RpcError as e:
            return False
        else:
            return True

    def HaveMessage2Client(self) -> bool:
        return self.__haveNewMessage

    def GetMessage2Client(self) -> Message2Clients.MessageToClient:
        self.__haveNewMessage = False
        return self.__message2Client

    def AddPlayer(self, playerID: int, playerType: THUAI6.PlayerType, humanType: THUAI6.HumanType, butcherType: THUAI6.ButcherType) -> None:
        def tMessage():
            try:
                playerMsg = THUAI62Proto.THUAI62ProtobufPlayer(
                    playerID, playerType, humanType, butcherType)
                for msg in self.__THUAI6Stub.AddPlayer(playerMsg):
                    self.__haveNewMessage = True
                    self.__message2Client = msg
            except grpc.RpcError as e:
                return

        threading.Thread(target=tMessage).start()
